pkg/kubelet/container/helpers.go

```diff
@@ -110,28 +110,20 @@ func ShouldContainerBeRestarted(container *v1.Container, pod *v1.Pod, podStatus
 // Note: remember to update hashValues in container_hash_test.go as well.
 func HashContainer(container *v1.Container) uint64 {
 	hash := fnv.New32a()
-	// Omit nil or empty field when calculating hash value
-	// Please see https://github.com/kubernetes/kubernetes/issues/53644
-	containerJSON, _ := json.Marshal(container)
+	containerJSON, _ := json.Marshal(pickFieldsToHash(container))
```

Comment:
A breaking change here.

This definitely changed the hash value of existing pods. When upgrading from lower version to v1.31, all pods would be restarted.

Reply:
Upgrading kubelet requires draining the pods on the node. In-place upgrade is not a community-supported upgrade method :)

ref: https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#drain-the-node

Reply:
@HirazawaUi This is a breaking change. No additional warning/message is given in the changelog/PR/Kubernetes docs. This is not acceptable!!!

Reply:
@dixudx Could you open an issue to discuss on the potential regression?

Reply:
Ah, since this is not a community-supported behavior, I didn’t include it in the release note.

I do know that some users in private cloud environments are using in-place upgrade solutions (although, as far as I know, public clouds typically drain pods when upgrading nodes). You’re more than welcome to open an issue so we can discuss this. :)

Reply:
We always assume node drains, and we do not spend much effort to make things like cgroups changes safe for in-place updates.

Reply:
Upgrading kubelets across minor versions without draining containers has never been supported and is explicitly not supported in documentation. We don't need to repeat all the things that aren't supported in every minor's release notes.

Reply:
Let's gather discussion in https://github.com/kubernetes/kubernetes/issues/129385 about whether to try to modify the current behavior in 1.31/1.32 or not.

I'd be strongly inclined not to, since backporting changes for this to patch releases of 1.31 and 1.32 risks breaking people who follow the documented guidelines for upgrading kubelets across minors, in an attempt to support in-place kubelet minor upgrades with running containers, which have never been supported.